lib/exercises.rb

```diff
@@ -1,29 +1,214 @@
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n log n)
+# Space Complexity: O(n)
+def top_k_frequent_elements_version2(list, k)
+  # edge cases
+  return list if list.empty? || list.size == 1
+
+  occurances_counter = {}
+  list.each do |digit|
+    if occurances_counter[digit]
+      occurances_counter[digit] += 1
+    else
+      occurances_counter[digit] = 1
+    end
+  end
+  # occurances_counter looks like this for input [1,1,2,2,2]: {1=>2, 2=>3}
+  # create an array out of this hash, sort it by values and return keys of k elements
+  sorted_occurances = occurances_counter.to_a.sort_by{ |pair| pair[1]}
+  number_pairs = sorted_occurances.slice(sorted_occurances.length - k, k)
+  result = number_pairs.map do |pair|
+    pair[0]
+  end
+  return result
+end
+
+# option 2 with time complexity O(n)
+# This method will return the k most common elements
+# in the case of a tie it will select the first occuring element.
+# Time Complexity: O(n)
+# Space Complexity: O(n)
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  # edge cases
+  return list if list.empty? || list.size == 1
+
+  occurances_counter = {}
+  list.each do |digit|
+    if occurances_counter[digit]
+      occurances_counter[digit] += 1
+    else
+      occurances_counter[digit] = 1
+    end
+  end
+  # occurances_counter looks like this for input [1,1,2,2,2]: {1=>2, 2=>3}
+
+  # create an array with the input number of elements + 1
+  # at indexes equal to occurance number I will store an array of occured digits
+  # +1 to have the last index available, e.g. [1,1,1] => [nil, nil, nil, (here at index 3 will be digit 1)]
+  sorted_occurances = Array.new(list.size + 1)
+  occurances_counter.each do |digit, counter|
+    if !sorted_occurances[counter]
+      sorted_occurances[counter] = Array.new
+      sorted_occurances[counter] << digit
+    else
+      sorted_occurances[counter] << digit
+    end
+  end
+  # sorted_occurances array for input c will be like: [nil, nil, [1,3], [2], nil, nil]
+  # reverse to keep the needed order of inner arrays elements, then flatten and get rid of nil elements
+  sorted_occurances = sorted_occurances.reverse.flatten.select { |element| !element.nil? }
+  # the result is: [2,3,1] where 1 is least frequent, 2 is most frequent in the input array
+  # the last step is to slice the k number of elements
+  result = sorted_occurances.slice(0, k)
+  return result
 end
 
 
+
+
+
+
 # This method will return the true if the table is still
 #   a valid sudoku table.
 #   Each element can either be a ".", or a digit 1-9
 #   The same digit cannot appear twice or more in the same 
 #   row, column or 3x3 subgrid
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n) where n is the number of all table elements
+# Space Complexity: O(n) - created 3 new storages
 def valid_sudoku(table)
```

Comment:
Since every Sudoku board is 9x9, Time complexity doesn't really apply.  😄 

Otherwise nice work!

Reply:
Oh true. It's going to be O(1). Thanks for catching that!